### PR TITLE
Adding minimal requestParameters schemas to all entry types

### PR DESCRIPTION
**Pull request by mbaudis**
As discussed in https://github.com/ga4gh-beacon/beacon-v2-Models/issues/50 it seems cleaner to have requestParameters defined for all entry types which at the minimum should contain the `id` parameter.

This PR add the documents as well as parameters for each `id` value defined in the schema (e.g. for `biosample` this would be `id` and `individualId`).

### CODE REVIEW

**Comment by Tom-Shorter** [comment]
This seems like a good start to the request parameters but I would like to see the ability to enter multiple ID's into fields, not just a single ID.

This does introduce the concept of an `OR` into the beacon spec though, I don't think this is a big issue as the `OR`s can only be between values of the same property and not between different properties.

Here's an example schema for dataset:

```
{
  "$schema": "https://raw.githubusercontent.com/ga4gh-beacon/beacon-framework-v2/main/requests/requestParameters.json",
  "dataset": {
    "type": "object",
    "properties": {
      "ids": {
        "type": "array",
        "items": {
          "type": "string"
        }
      }
    }
  }
}
```

and an example partial request:

```
{
  "dataset": {
    "ids": [
      "dataset1",
      "dataset2",
      "dataset3"
    ]
  }
}
```
This would return information about the 3 datasets, this would be equivalent to someone sending in 3 requests with 1 dataset ID in each or the user going to /datasets/{id} for each dataset.

I would also suggest that we pluralise the "id" property to be "ids"

**Comment by mbaudis** [comment]
@Tom-Shorter Good point & we actually support this - e.g. the query demonstrating the VRS does exactly this...

http://progenetix.org/beacon/variants/?variantIds=pgxvar-5c86664409d374f2dc4eeb93,pgxvar-5be1840772798347f0eda0d8 

Regarding OR: This still keeps it away from filters (though we had discussions about this before...); so one could be explicit for "**document retrieval by ids**", not really a "search".

File: BEACON-V2-Model/biosamples/requestParameters.json [hunk 1]
```diff
@@ -3,12 +3,18 @@
...
     "properties": {
-      "id": {
-        "type": "string"
+      "ids": {
+        "type": "array",
+        "items": {
+          "type": "string"
+        },
+      },
+      "individualIds": {
```
Comment (Tom-Shorter) [inline BEACON-V2-Model/biosamples/requestParameters.json:12]: I'm not sure this is required here. If you want to make a query which specifies individualIDs you'd send the individuals model as part of the query and specify the IDs within there.

File: BEACON-V2-Model/biosamples/endpoints.json [hunk 2]
```diff
@@ -305,6 +307,26 @@
...
+        "name": "id",
+        "in": "query",
+        "schema": {
+          "type": "array",
+          "items": {
+            "type": "string"
+          },
+        },
+      },
+      "individualIds": {
```
Comment (Tom-Shorter) [inline BEACON-V2-Model/biosamples/endpoints.json:320]: Whilst this is correct as a potential GET parameter we should probably move this out to be defined within the individuals model, this GET parameter will only be available if individuals is supported by the beacon. Each entryType will come with a list of possible GET parameters that can be used with all query endpoints so packaging the GET parameters with the entry type seems cleaner and allows much easier creation of custom models + GET parameters

**Comment by mbaudis** [comment]
@Tom-Shorter So to keep it simple we just should have the option of multiple `ids` per entry type & then go over `requestedSchema`, if alternative deliveries are supported? I guess this would be good enough for the main data entities.

**Comment by Tom-Shorter** [comment]
@mbaudis I'm not sure where the `requestedSchema` comes into play here? Although `requestedSchemas` is a part of the request body and determines the response type it has no direct bearing on the query itself.

If I get a chance today I'll mock up what I believe a requestParameters.json file should look like and an example request

**Comment by mbaudis** [comment]
@Tom-Shorter The requestedSchema allows you to e.g. get the biosamples from an individuals query. For a path-based request this is `/individuals/{id}/biosamples`; i.e. retrieve the biosamples associated with a single individual. If one now generally uses `ids` as a query parameter for each entry type one can retrieve e.g. all biosamples for several individuals through `/individuals?ids=ind1,ind2,ind3&requestedSchema=biosample` - instead of defining `individualIds` as request parameter for the `biosamples` entry type.

IMO.

**Comment by Tom-Shorter** [comment]
I hadn't thought about that use case of `requestedSchemas`, in my mind the response will always be of the right most entry type within the URL, this would be individuals for `/individuals` and `/biosamples/{id}/individuals` or biosamples for `/biosamples` and `/individuals/{id}/biosamples` , the `requestedSchema` allowed you to then choose between the default schema or a custom schema which is also made available in that beacon, I hadn't thought about the ability to enter a different entry type than what's defined within the URL.

If `requestedSchemas` can be used to dictate the response type then I'm wondering why we need all of the different query endpoints when instead we could send a query to a single endpoint and use `requestedSchemas` to define the response entry type.

I'm guessing that I've made an incorrect assumption about how something behaves somewhere along the way!

**Comment by mbaudis** [comment]
@Tom-Shorter that's all correct; but the definition by path _only_ allows to retrieve the "rightmost" schema's documents for a _single_ "leftmost" entry type. In principle the path compositions wouldn't be needed if having `id(s)` queries; but they offer a standardized access to single data points.

**Comment by mbaudis** [comment]
@colinveal `/individuals/{id}/biosamples` etc. - which has been in the endpoints forever - retrieves one or more biosamples associated with this individual, if the endpoint pattern is defined; this seems pretty clear to me (and I'm not responsible here ...). And the schema used in the response is the default schema for the rightmost entry type; _i.e._ `/biosamples/defaultSchema.json`. This doesn't seem controversial?

Obviously, the endpoint pattern `/individuals/{id}/biosamples` has to be defined - if the "biosamples from individuals" response isn't supported this pattern shouldn't exist. And you could have other patterns, if providing a schema for them - e.g. `/individuals/{id}/phenopackets`, `/individuals/{id}/FHIR` if they are part of your model. But ask @jrambla ...

How the query is handled internally is a thing for the implementers. The default data models have `individualId` in the `biosample` schema.

My original point _was_ the definition of `individualIds` as parameter for `/biosamples` queries, _i.e._ an explicit specification. But the single `id` per path & different response schema already exists.

Summary:
* current REST patterns allow the response with a defined schema different from the entry type
* the paths only allow retrieval of documents associated with a single entry type document identified by its `id` (e.g. all `genomicVariations` of a `biosamples` document)
* the current default models need a definition of query parameters, including assertions about optional plurality (`ids` vs. `id` etc.)
* while one could define requestParameters corresponding to `id`-type properties of individual schemas (e.g. `individualIds`) one could also limit the pre-definition to the `id(s)` as they correspond to the current scenario of path-based retrievals (see @Tom-Shorter 's comment https://github.com/ga4gh-beacon/beacon-v2-Models/pull/112/files/2e2bc4da0e996774d73f1b1471ca21f1c5a94391#r841873244) - but IMO with _pluralization_

**Comment by jrambla** [comment]
The `id` of the entryType neither of the related entry Types MUST NOT GO in the parameters.
The correct way to do that filtering is by using the REST path.
The only ones that we agreed on adding were the collection ones if we prefer:
`biosamples?datasetIds=1,2,3`
to
```
dataset/1/biosamples
dataset/2/biosamples
dataset/3/biosamples
```
Anything else is plainly against REST, confusing and prone to inconsistencies between the URL path and the requestParameters.

**Comment by jrambla** [comment]
`requestedSchema` is just for returning the model for the rightmost entryTypo NOT for changing the entryType returned.

**Comment by jrambla** [comment]
Something like `/individuals/1/FHIR` is also breaking the REST principal.
REST is  based on "Resources" (entryTypes in Beacon lingo), and `FHIR`, neither `phenopackets `is a resource but a model for a resource type. These are the ones that should go into requestSchema.
`/individuals/1?requestedSchema=FHIR` or `/individuals/1?requestedSchema=phenopackets`

**Review by jrambla — COMMENTED** [review]
None of the changes suggested here are following the REST principles.
Therefore, I fear, we MUST reject all of them.

**Comment by mbaudis** [comment]
@jrambla Sure the `/individuals/1?requestedSchema=phenopackets` part is clear. But then I don't get what `/individuals/1/genomicVariations` would be about?

**Comment by jrambla** [comment]
`/individuals/1/genomicVariations` is returning all the `genomicVariations` of individual 1

**Comment by mbaudis** [comment]
So mI had this right. But `requestedSchema` is the `Schema to be used to format the `result` field in the response`. So `/individuals/1?requestedSchema=genomicVariations` would respond as above? Or would it be a conflict?

**Comment by jrambla** [comment]
`/individuals/1?requestedSchema=genomicVariations` should "return an error" as this path & query string are requesting a given entryType (i.e.: `individuals`) but asking for an schema that applies to another entryType (`genomicVariations` is indeed not an schema but an entryType name, but let's stick to it for this conversation).

To know which requestedSchema strings are valid for each entryType, the Beacon client MUST use the informational endpoints.
In particular:
` /configuration`, that would return the `beaconConfiguration.json` contents, where the entryType names are associated with the `additionallySupportedSchemas`, which lists the alternative schemas supported by that specific Beacon instance.

```
    "individual": {
      "id": "individual",
      "name": "Individual",
...
      "defaultSchema": {
        "id": "ga4gh-beacon-individual-v2.0.0",
        "name": "Default schema for an individual",
        "referenceToSchemaDefinition": "https://exampleBeacons.org/individuals/defaultSchema.json",
        "schemaVersion": "v2.0.0"
      },
      **"additionallySupportedSchemas": []**
    },
```


For the client to find out which are the supported URL paths, it MUST call 
` /map`, that would return the `beaconMap.json` contents, where the entryType endpoints and related entryTypes are listed.
e.g. 
```
    "analysis": {
      "entryType": "analysis",
      "openAPIEndpointsDefinition": "https://exampleBeacons.org/analyses/endpoints.json",
      "rootUrl": "https://exampleBeacons.org/analyses",
      "singleEntryUrl": "https://exampleBeacons.org/analyses/{id}",
      "endpoints": {
        "genomicVariant": {
          "returnedEntryType": "genomicVariant",
          "url": "https://exampleBeacons.org/analyses/{id}/g_variants"
        }        
      }
    }
```

**Comment by jrambla** [comment]
@Tom-Shorter there is a difference between `ids` and `filters`.
Filters are "static" values that many entries could have (e.g, you can have N females and M males in your cohort), while ids point to a single entry (e.g. individual #1).

We decided not to go for a detailed list of requestParameter for every attribute in any entryType, because that would be overkill and actually not necessary. You can have the same result simply qualifying everything as you are doing inside your parameters definition. So, as a backend you must be able to parse `individual.sex.id` as it defines a path inside the model. However, the expectation is that the backend is not storing data in exactly the same format than the model. Hence, the backend would eventually need to translate `individual.sex.id` into table 'People', column 'gender', value 'M'. We believed that this is a similar effort than translating `PATO:000383` into that same internal path.

In other words, we favoured using `filters=PATO:000383` versus `individual.sex.id=female` as the ontology term is like a condensed expression which includes the attribute and the value.

Hope this clarifies

**Comment by mbaudis** [comment]
Exactly. I don't know where exactly, but we also should emphasise that the Beacon default model:

* has no explicit relation to how a local data organisation / model is supposed to look
* _can_ serve as guidance for assumptions (e.g. if you represent data from biosamples and variants then there is a one-to-many relation between those)
* provides schemas for how to represent your data in responses
* 
So Beacon is mostly agnostic about where/how you store the value for `sex`. It just relies on that you expose `filters` and if they contain a value corresponding to `sex` that this can be used to query/filter your data (so that the response results all _filtered_ by it). 

Btw.: While I do not see much of a technical difference between `filters` and `queryParameters` when doing an implementation, _logically_ there is the difference between tightly scoped, model-dependent query parameters and "post-model" filters.

**Comment by mbaudis** [comment]
@jrambla Still - we need a way to query lists of `ids`, not only for datasets (IMO but others seem to concur). So why not having `ids` (plural) for a given entry type, if it is **not** specified in the path? E.g. only in `/biosamples`, but not in `/biosamples/{id}`.

**Comment by jrambla** [comment]
I would like to have a clear use case that we couldn't/wouldn't solve with multiple queries...
OpenAPI allows for multiple Ids in the path, but honestly I haven't seen that implemented anywhere or I haven't noticed it.

**Comment by mbaudis** [comment]
@jrambla 
* main reason is to recapitulate functionality of REST paths through query - e.g. form-based requests by id(s)
* combination of several entries
    - many reasons to be envisioned - e.g. 2-case comparisons, 2/more variants, ref vs. case...
    - see e.g. example we submitted in the PRC response (I've cleaned the URL to match the proposed definitions - works...) <https://progenetix.org/beacon/variants?ids=pgxvar-5c86664409d374f2dc4eeb93,pgxvar-5be1840772798347f0eda0d8>
* we allow it for datasets

In principle we could do this over filters, but then:

* doesn't work w/ AND paradigm
* specific "id filter" prefix would need to be defined ...

But then you also said:

> The only ones that we agreed on adding were the collection ones if we prefer:
`biosamples?datasetIds=1,2,3`
> to
> ```
> dataset/1/biosamples
> dataset/2/biosamples
> dataset/3/biosamples
> ```

... which AFAIK isn't defined anywhere! It only appears in one framework example (we have it in the Progenetix stack, but it is _not_ in the spec?!).

**Comment by Tom-Shorter** [comment]
@jrambla it's not that multiple queries couldn't/wouldn't solve the problem, its that multiple queries is precisely that, multiple queries. Sending multiple IDs within a POST request is extremely simple to implement as both a user writing the query and as a beacon owner interpreting the query and can drastically improve performance and give a better understanding of the returned data to the user as the returned counts will be aggregated across all query IDs, not just one query ID at a time. This does not stop queries from being sent with single IDs either.

I don't agree with sending multiple ID's within the GET path, it's not common and honestly looks messy. If you need to provide multiple IDs then use either a GET query param or the POST request body.

My argument for multiple IDs:

Imagine wanting to know if a beacon has biosamples (and an estimate on the number of biosamples) related to a set of rsids (a few 100 in length). I do not imagine the researcher making the queries by hand, instead they would be using a UI where they would likely be able to copy and paste their list of rsids and the UI handles all the query generation.

If the UI has to make a request for each rsid then it would need to send out 100s of queries to the beacon, this would take a long time to finish (probably in the region of minutes depending on the beacon backend) and the results would come back one by one for each rsid. The user then has no idea what level of replication there is within the counts, if rsid1 return 20 biosamples and rsid2 return 15 then what level of overlap is there between the 2, are the 15 all included within the 20 or are the 15 completely different to the first 20? The user has no way of finding this information out and it could make a big difference, in this example the user knows there's somewhere between 20 and 35 biosamples they might be interested in but then add in the other few 100 results and this range would become almost meaningless, the range would have to be guessed at somewhere between the highest number of biosamples returned from one query all the way up to every count for every query added summed, something like 20 - 6000 for a few 100 queries each returning 20-30 results. 

Had the rsids been sent as one query the researcher would find the count of distinct biosamples which match any of the rsids, maybe something like a count of 60 would be returned in this case as there is a huge amount of overlap. The user now has a more exact figure to work from, something which is not possible to obtain from sending multiple queries. The query would also likely return sooner than sending 100s of queries out and waiting for the final one, checking if a value is within a list of values is pretty quick and although slower than exact matches it will definitely out perform having to do 100s of individual exact matches. This is also ignoring that for every query that comes in the beacon will need to validate the request, translate the request into a query that can be used on their data, read the response from the DB and translate it into a beacon response.

Overall I cannot see any reason why we shouldn't do multiple IDs in a query, it takes nothing away from the current functionality but makes new functionality available.

Also I understand that RSIDs might not be the same as a datasetID in this regard but the same principles still apply if looking for biosamples related to a list of 300 individuals for example. * edit: individuals isn't an exact replacemnent for rsids here due to the relationship between individuals and biosamples (one - many) instead of what biosamples and variants have (many - many), variants is also a special case that has defined request parameters but something like the rsid doesn't seem to be one, and if it is then it's not an array.

**Comment by mbaudis** [comment]
> I don't agree with sending multiple ID's within the GET path, it's not common and honestly looks messy. If you need to provide multiple IDs then use either a GET query param or the POST request body.

and 

> Overall I cannot see any reason why we shouldn't do multiple IDs in a query, it takes nothing away from the current functionality but makes new functionality available.

Both +1 (with support for multiple IDs in GET but not in the path).

**Comment by jrambla** [comment]
This is a productive discussion, and don't get my replies as me being against the ideas described above, just exploring how they impact what we have now:

**Consideration:** We must distinguish between ids of related stuff vs, ids of the leftmost entryType. Hence we must distinguish between:
1) `/biosamples/1,2,3/g_variants` that must return all the genomicVariations of biosample 1, 2 and 3
from 
2) `/g_variants?geneIds:BRAF,TP53` which will return all genomicVariations that are having such values in such attribute. Although we call them `id`, they are just strings inside one attribute.
When I'm referring to multiple ids, I'm referring to the first example. The second example is already possible provided that the parameter is defined, and defined as an array (which is not in the case for `geneId`). This is the suggestion for the datasetIds.

Assuming that we have a request parameter that allows to query several geneIds, which would be the procedure for getting the biosamples that contain such genomicVariations?
- The user (through the UI) would do the query (1) `/g_variants?geneIds:BRAF,TP53` which will return all genomicVariation details, > which include the biosampleIds inside the `caseLevelData/biosampleId` attribute. 
- About the biosamples, this query is only returning their ids and nothing else.

At this point is where we must decide how to proceed:
1. Should the UI query each biosample to be able to show some descriptive attributes?
2. Should the user click one biosample at a time?
3. Should the user select some of them and then the UI getting just these samples?
4. Should the user to a query that filters biosamples according to some other attributes and, in the background, the UI intersecting the biosample ids between query 1 and query 2?
5. Should the user to a query that filters biosamples according to some other attributes and adding all the biosample ids from query 1 as multiple Ids like in example 1?
(The later would be something like `/biosamples?ids:bs1,bs4,bs9&sampleOriginDetail:blood`)

In the shake of simplicity in the spec and implementation of the backend (not the UI), we have been considering the option 4 is enough for Beacon version 2, waiting for feedback on actual use and how to evolve the Model. The discussion above is in favor of option 5.
Are we agreeing on this view? 
Are we preferring option 5 to option 4?

**Comment by Tom-Shorter** [comment]
> Assuming that we have a request parameter that allows to query several geneIds, which would be the procedure for getting the biosamples that contain such genomicVariations?

This is where I might be considering queries differently to you. In this instance I would not go to the /g_variants endpoint, my end goal is biosamples so I would want to query /biosamples directly and basically carry out the example you put in option 5 but replace ids with geneIDs. 

I can easily envision people wanting to send queries in the same manner as your example and it would likely be a common approach as well. When querying anything myself I often check if the resource has information relating to what I'm going to query about (geneIDs in this case) and if they do I will then perform the actual query I'm interested in based off of the results of the first query, usually by extending the query or sending it to another endpoint. In the case of beacon I wouldn't need to use the results from the first query (unless I want to add/remove responses) to form the second query as it seems more beacon like to send the same gvariants query as used in query1 to a different endpoint to get the actual results I want. 

There is an interesting use case for filters here, for example in the first query if I found a group of variants within the results I wasn't interested in I could extend the query to add a filter or filters to exclude these from the results. 

I was always under the impression that the beacon models are to be used for both the response and the request and any number of models could be sent to any endpoint. The models would be sent via the requestParameters section and would be used to query any fields which are found within the model. With the current approach (requestParameters are just for IDs and basically everything else is a filter) the idea of using a model as part of the request no longer exists, or at least not in the same way as I was thinking. On a side note, using my approach there would be a massive overlap between filters and requestParameters so my assumption for filters was that they were to be used more for data that lies outside of the models, these could be additional data points a beacon instance holds and although not accessible through the standard model they would be accessible through the filters. This assumption would mean that my proposed filters use case to exclude responses wouldn't work if I wanted to exclude results based on data points within the models.

I also want to check genomic variations, are these a special use case where you use the request parameters found within requestParameters.json and send these via the requestParameters section? If this is the case then what determines whether a model should be treated like gvariants and passed through as requestParameters or treated like most models and the IDs are the main requestParameters? If this isn't clearly defined then, to me, it looks like an inconsistency within the beacon model which could open the way for people to exploit this and use the genomicVariants as a template for all their models and define the requestParameters to be used instead of using filters, this option would seem appealing to networks especially as it allows the request parameters to cover the core data points shared by all networks and the filters to cover non-common data points.

tldr; option 5 is closest to what I was envisioning with beacon but my understanding of how the current beacon works is not clear due to these recent discussions

**Comment by jrambla** [comment]
I see what you are saying Tom. We are dealing with different needs/requests/considerations here:
Remember the principal of making things simple and uniform enough between Beacon instances.
If we foster in the model using attributes in one entryType (p.ej. geneId from genomicVariations) in another entryType (e.g. biosamples) the number of combinations could be huge. Then we should face the question of "should we make these requestParameters mandatory and make implementations harder/longer? or should we keep them optional and the we can't rely on all Beacon instances having them implemented?"
In the current design, we opted for the 2nd one, and to complement that by making the Beacon instance declarative on what requestParameters it is supporting. The double query approach gives us the same result w/o having such dilemma.
Given that optional nature of the request parameters, in the initial version of the model we left out almost all of them. We are enriching it as we are agreeing that some are "badly" needed or could be of big help.

The second point is about Query By Example (QBE) using the Model as query for a Beacon instance. This is something we discussed initially (if my memory is not tricking me) with Tony and decided to keep it for a later version. IMHO, QBE is significantly more complex than sending one Model and getting back the result as you'll probably want multiple values, negations and so on, that are not part of the dry Model. It is an interesting approach, suggested also by the Phenopackets team, but requiring some more work before we can add it to Beacon.

The role of filters is a complement to the request parameters (or viceversa). As explained elsewhere, request parameters should be used for things that have IDs o ID-like nature, while filters are used for attributes or indeed numeric expressions and fuzzy comparisons. One exception is the genomicVariations one, mainly to address expectations from Beacon v1 implementers and users that could be assuming that a Beacon v1-like query would be possible in Beacon v2... hence we offer them parameters that allow building such type of query... otherwise we would rely on filters (probably).

In our conversations there is always another consideration: "We aimed for a simple model and framework that would 'react' to the feedback we get from the initial round of users. We tried to avoid being too much prescriptive as this is making the model less simple"... The very first conversations we had about Beacon v2 were stating that they want a Bv1 with more context and a bunch of filter... I fear that if we depart too much from that we could be losing that simplicity.

**Comment by Tom-Shorter** [comment]
Thanks for the clarification @jrambla, this has helped a lot and I can see how the current beacon works. I was looking at it from a position where I had been present in the introduction of some of the principals but then not when it was decided they would be retained for future versions. I'll have to have a go over the spec and docs in some more detail now to try and fully understand the changes I've missed.

**Comment by colinveal** [comment]
@jrambla  The multistage query using IDs you propose above only works if IDs are returned. How are more in depth queries, say spanning different models, work if the implementation only responds in counts or boolean and not IDs?
